creator-node/src/routes/files.js

```diff
@@ -240,115 +259,220 @@ const getCID = async (req, res) => {
+        stage: `FS_STATS_LEGACY_STORAGE_PATH`,
         time: `${Date.now() - startMs}ms`
       })
 
-      if (!queryResults) {
+      if (fsStats.isFile()) {
+        decisionTree.push({
+          stage: `CID_CONFIRMED_FILE_LEGACY_STORAGE_PATH`
+        })
+        fileFoundOnFS = true
+      } else if (fsStats.isDirectory()) {
         decisionTree.push({
-          stage: `DB_CID_QUERY_CID_NOT_FOUND`
+          stage: `CID_CONFIRMED_DIRECTORY_LEGACY_STORAGE_PATH`
         })
         logGetCIDDecisionTree(decisionTree, req)
         return sendResponse(
           req,
           res,
-          errorResponseNotFound(
-            `${logPrefix} No valid file found for provided CID`
-          )
+          errorResponseBadRequest('this dag node is a directory')
         )
-      } else if (queryResults.type === 'dir') {
+      } else {
         decisionTree.push({
-          stage: `DB_CID_QUERY_CONFIRMED_DIR`
+          stage: `CID_INVALID_TYPE_LEGACY_STORAGE_PATH`
         })
         logGetCIDDecisionTree(decisionTree, req)
         return sendResponse(
           req,
           res,
-          errorResponseBadRequest('this dag node is a directory')
+          errorResponseBadRequest('CID is of invalid file type')
         )
-      } else {
-        decisionTree.push({
-          stage: `DB_CID_QUERY_CID_FOUND`
-        })
-        storagePath = queryResults.storagePath
       }
     } catch (e) {
       decisionTree.push({
-        stage: `DB_CID_QUERY_ERROR`,
+        stage: `FS_STATS_LEGACY_STORAGE_PATH_CID_NOT_FOUND`,
+        time: `${Date.now() - startMs}ms`
+      })
+      // continue
+    }
+
+    if (fileFoundOnFS) {
+      storagePath = legacyStoragePath
+    }
+  }
+
+  // If client has provided filename, set filename in header to be auto-populated in download prompt.
+  if (req.query.filename) {
+    res.setHeader('Content-Disposition', contentDisposition(req.query.filename))
+  }
+
+  // Set the CID cache-control so that client caches the response for 30 days
+  res.setHeader('cache-control', 'public, max-age=2592000, immutable')
+
+  /**
+   * If file found on file system:
+   * 1. add background task to rehydrate from disk to IPFS
+   * 2. stream from FS
+   */
+  if (fileFoundOnFS) {
+    /**
+     * 1. add background task to rehydrate from disk to IPFS
+     */
+    startMs = Date.now()
+    try {
+      RehydrateIpfsQueue.addRehydrateIpfsFromFsIfNecessaryTask(
+        CID,
+        storagePath,
+        {
+          logContext: req.logContext
+        }
+      )
+      decisionTree.push({
+        stage: `ADD_REHYDRATE_IPFS_FROM_FS`,
+        time: `${Date.now() - startMs}ms`
+      })
+    } catch (e) {
+      decisionTree.push({
+        stage: `ADD_REHYDRATE_IPFS_FROM_ERROR`,
         time: `${Date.now() - startMs}ms`,
         error: `${e.message}`
       })
+    }
+
+    /**
+     * 2. stream from FS
+     */
+    startMs = Date.now()
+    try {
+      const fsStream = await streamFromFileSystem(
```

Comment:
i know you didn't add this code but i'm curious if you know. streamFromFileSystem https://github.com/AudiusProject/audius-protocol/blob/babeaa0dee50eb86c1220f52b60c9930c5f24254/creator-node/src/routes/files.js#L118-L128 does a pipe and res.end(). i’m not sure what setting as variable does

Reply:
good question, previously the call was `return await streamFromFileSystem` so i just separated that. clearly it works
i'll dig in later tn

Reply:
discussed offline - this is not causing any issues since function always sends response to client, the return statement is most likely just redundant. i'll clear this up in a fast follow refactor